src/Record.js

```diff
@@ -33,6 +33,12 @@ export class Record {
   constructor(defaultValues, name) {
     let hasInitialized;
 
+    if (isRecord(defaultValues)) {
+      throw new Error(
```

Comment:
@conartist6  The fact is that `isCollection` and `isKeyed` returns `false` as the implementation does check the existence of `@@__IMMUTABLE_ITERABLE__@@` and `@@__IMMUTABLE_KEYED__@@`, which are undefined for Record instances.

I changed the code with `toObject` as you suggested to simply fix the original issue, as it seems to dangerous to check the default value in the `Record` method (for now at least as it may induce bugs for weird objects).

The fact is, that now you can not pass a `Map` as defaultValues (but it wasn't possible either).

As a matter of fact, the record defaultValues can be really weird, as it accepts everything where you can call `Object.keys` on (even if the typescript types only allow instances of `Object`).

Maybe I should still add an error like this 
```js
if (typeof defaultValues !== 'object' || defaultValues === null) {
  throw new Error(
    'Can not call `Record` with a non-object as default values. You may want to use a javascript object.'
  );
}
```
and then later check if it's has the `toObject` or `toJS` method as default values ? 🤔 

What do you think ?

Reply:
My point generally is this: we should prefer working over throwing an error when both are possible. I forgot that records are not collections, but it doesn't change my general thought. In your case of course you didn't mean to pass a record, but think about somebody who did mean to do that. If we can understand what they meant, why should we not just do it?

Reply:
I am fine with throwing on collections. But for the same reasons you took this up initially it would be nice to tell the user what they did wrong. Of course they wouldn't be in an infinite loop, but the results would definitely be surprising.

Reply:
I do not know what is the direction of immutable about this point. The two options are valid: either throw to avoid unexpected comportment or try to correct user code whenever possible. 🤷‍♂️ 

Apparently immutable 3.8 does not address neither one or another solution on this specific feature.

I finally did two things:
- accept `Record` as a constructor
- throw when the default values are not an object

This way, the 3.8 comportment still works as expected, and there is an explicit error message when we try to inject weird default values.

We could eventually manage some other selected types (like `Map`, which is really similar to a JS object)

Reply:
I don't think it's safe to assume what the author intended in this case. For example, the author of the original issue identified that using `record.toObject()` was actually [not what he was intending](https://github.com/immutable-js/immutable-js/issues/1690#issuecomment-472546059) and it was actually a bug in his code.

I agree that we should very rarely throw `Error`s, but since I don't think we know the author's intention, and we know it's going to throw a more ambiguous `Error` later on, we might as well throw an error to let the author know what's happening, and then they can easily identify the problem and figure out what to do.

I would prefer that we only throw an `Error` if it's a `Record` and not worry about other cases for the moment. Immutable very rarely validates its inputs (from what I can tell), this is just a special case because we know that it's going to throw an Error, and so we might as well provide a helpful error message explaining what's happening.

Reply:
My point of view is that we should throw on all Immutable instances and on every thing that is not an acceptable input as one of the first thing that the `Record()` function will do is to call [`Object.keys`](https://github.com/immutable-js-oss/immutable-js/blob/master/src/Record.js#L45), and that will throw the unintelligible error 

With a immutable Map:
 
> TypeError: Cannot read property 'get' of undefined
> 
> _at Record.get (dist/immutable.js:5577:30)_
>       _at Record.get (dist/immutable.js:5688:23)_
>       _at Record.set (dist/immutable.js:5691:26)_
>       _at new Record (dist/immutable.js:5523:22)_
>       _at Record (dist/immutable.js:5491:16)_

 
With null:
>     TypeError: Cannot convert undefined or null to object
>      at Function.keys (<anonymous>)

With the string `abc`, no error but a factory that seems really unwanted:

> { '0': 'a', '1': 'b', '2': 'c' }

With the array [123, 'abc']

> { '0': 123, '1': 'abc' }

So I think an understandable error is way more powerfull than anything that will throw afterward.

Reply:
Okay, I did a little bit more digging to see what the other Immutable data structures do. Many of them end up calling `keyedSeqFromValue`, `indexedSeqFromValue`, or `seqFromValue` ([code](https://github.com/immutable-js/immutable-js/blob/master/src/Seq.js#L296-L335)) which throws a `TypeError` if they're not of a proper type. Based on that I'm fine with throwing an error on immutable instances and non-objects.

I would still prefer not implicitly converting the `Record` since I don't think it's clear what the author's intention is.

Thanks for sticking through these long conversations!

Reply:
Done

Reply:
I'm fine this, but I do disagree with @Methuselah96. Yes this one person passed a record and it wasn't what he meant. But the purpose of records is to behave like objects. There happens to be this one flaw in their facade as regards `Object.keys`, but I don't think we want to save that guy from himself. If the program can be executed we can't fix programmer error at runtime... Typescript probably could have caught the problem though.

Reply:
I'm going to further defend my position, but feel free to stop me when this has become counter-productive.

I can't think of a case where I would want to use a `Record` as the `defaultValues` for a new `Record` in any code I've written, and I would rather get an error (and probably uncover a bug) and be forced to call `.toObject()` on the `Record` to make my intention clear (if it wasn't a bug). I would prefer to be saved from myself in this scenario.

I do agree that records are to behave like objects, but they're more heavy-weight because you have to define the structure of the "object" before you can actually create the "object." That's why I have a hard time believing that someone would choose to use this more heavy-weight "object" in order to just use as default values for another heavy-weight "object."

Reply:
Yeah I agree with you that a legitimate use is pretty unlikely. I'm giving my approval to what @jdeniau has got now.